.github/sync_labels.py

```diff
@@ -0,0 +1,348 @@
+            # pull down menue). In this case `label` has been removed
+            # on the `on_label_add` of the first of the two labels
+            partn = self.active_partners(label)
+            if partn:
+                self.add_comment('Label *%s* can not be added due to *%s*!' % (label, partn[0]))
+            else:
+                warning('Label %s of %s not found!' % (label, self._issue))
+            return
+
+        item = sel_list(label)
+        if sel_list is State and self._pr:
+            if not self.is_pull_request():
+                if item != State.needs_info:
+                    self.add_comment('Label *%s* can not be added to an issue. Please use it on the corresponding PR' % label)
+                    self.remove_label(label)
+                    return
+
+            if item in [State.positive_review, State.needs_work]:
+                self.add_comment('Label *%s* can not be added. Please use the corresponding functionality of GitHub' % label)
+                self.remove_label(label)
+                return
+
+        for other in sel_list:
+            if other != item:
+                self.remove_label(other.value)
+
+    def select_label(self, label):
+        """
+        Add the given label and remove all others.
+        """
+        self.add_label(label)
+        self.on_label_add(label)
+
+    def remove_label(self, label):
+        """
+        Remove the given label from the issue or PR of the handler.
+        """
+        if label in self.get_labels():
+            self.edit(label, '--remove-label')
+            info('Remove label from %s: %s' % (self._issue, label))
+
+    def on_label_remove(self, label):
+        """
+        Check if the given label belongs to a selection list. If so, reject
+        the removement in case it represents the stat of a PR.  In all other
+        cases add the successor of the label except if there is none or there
+        exists another label of the corresponding list.
+        """
+        sel_list = selection_list(label)
+        if not sel_list:
+            return
+
+        if label in self.get_labels():
+            # this is possible if two labels of the same selection list
+            # have been removed in one step (via multiple selection in the
+            # pull down menue). In this case `label` has been added
+            # on the `on_label_remove` of the first of the two labels.
+            partn = self.active_partners(label)
+            if not partn:
+                self.add_comment('Label *%s* can not be removed (last one of list)!' % label)
+            else:
+                self.on_label_add(partn[0])
+            return
+
+        item = sel_list(label)
+        if sel_list is State and self._pr:
+            if item in [State.positive_review, State.needs_work]:
+                self.add_comment('Label *%s* can not be removed. Please use the corresponding functionality of GitHub' % label)
```

Comment:
I would still allow people to actively overwrite the label. For example, if the PR creator realizes she still needs to fix something small, then she can simply add a "needs work" label (but you cannot "request changes" on your own PR). Similarly, not every positive review means that it can directly be merged (e.g. second pairs of eyes needed).

Reply:
I agree. Perhaps we should not think "positive review" == "approval". Adding "positive review" label is one person's opinion. She may be another person who gives approval to the PR.

Reply:
> I would still allow people to actively overwrite the label. For example, if the PR creator realizes she still needs to fix something small, then she can simply add a "needs work" label (but you cannot "request changes" on your own PR). Similarly, not every positive review means that it can directly be merged (e.g. second pairs of eyes needed).

I agree with respect to `needs work`. Your second point I don't understand: How has *second pairs of eyes needed* be realized in Trac other than selecting `positive review` at the end by the second pairs of eyes?

Reply:
> I agree. Perhaps we should not think "positive review" == "approval". Adding "positive review" label is one person's opinion. She may be another person who gives approval to the PR.

As I said above: I think the meaning of `positive review` should be as closed to the meaning we are used to from Trac. If we leave the meaning to *one person's opinion* then the meaning is unknown and the label is not only useless but confusing. Then, as suggested before, we better should remove it.

Reply:
On trac, you can add yourself as reviewer and comment that you agree with the changes, but don't set the ticket to "positive review". Something like this should still be possible on github, i.e. you review a PR positively but don't add the "positive review" label (or remove it if it has been automatically added).

Reply:
Perhaps we can refine the check on adding or removing a label using the GitHub context `github.actor` depending whether he matches the author of the PR or corresponding reviews or whether he has extra permissions (i.e. belongs to the Triage team, if this can be checked).

Reply:
I always allow to add the `needs_work` label. This adds a review of `github-action` requesting changes. Adding `positive_review` I allow in the following cases (not sure if they are strong enought and cover what you meant by second pairs of eyes):

1. There must be at least one review of a member
2. There should not be changes requested by some one else
3. If in addition the actor is different from the author it is allowed
3. Elsewise, more checks are needed:
  a. There must be at least one review of someone else
  b. There must be at least one commit of someone else

If all conditions are satisfied the PR will be approved by `github-action`. Else the label is removed and a comment: *Label can not be added. Please use the corresponding functionality of GitHub* is posted.

Reply:
That sounds very reasonable! Concerning "There should not be changes requested by some one else" not sure what happens in the case someone is against the changes but multiple other people give their positive review. But maybe this situation is rare enough to not need special treatment?

Reply:
What should be possible is: after giving a positive review, remove the label "s: positive review" and add instead the "needs review" label (meaning "I agree with the changes, but someone else should also have a look at this PR").

Reply:
> That sounds very reasonable! Concerning "There should not be changes requested by some one else" not sure what happens in the case someone is against the changes but multiple other people give their positive review. But maybe this situation is rare enough to not need special treatment?

I think such a conflict should be resolved using the GitHub review functionality, since comments on the respective review that request changes are needed.

Reply:
> What should be possible is: after giving a positive review, remove the label "s: positive review" and add instead the "needs review" label (meaning "I agree with the changes, but someone else should also have a look at this PR").

You can always add `s: needs review` and  `s: needs info`. In both cases the `s: positive review` will be removed by the bot.